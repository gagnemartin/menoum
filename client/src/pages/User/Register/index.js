import { useState } from 'react'
import { useHistory, useLocation } from 'react-router-dom'
import { useUserDispatch } from '../../../hooks/useUser'
import { isSuccessResponse } from '../../../global/helpers'
import useFormInput from '../../../hooks/useFormInput'
import { actionTypes } from '../../../reducers/userReducer'
import { UsersService } from '../../../services'

const Register = () => {
  const history = useHistory()
  const location = useLocation()
  const dispatch = useUserDispatch()

  const email = useFormInput('')
  const password = useFormInput('')
  const confirmPassword = useFormInput('')

  const [errors, setErrors] = useState({})
  const [isLoading, setIsLoading] = useState(false)

  const register = async (data) => {
    dispatch({ type: actionTypes.request, loading: true })
    try {
      const payload = await UsersService.register(data)

      if (isSuccessResponse(payload)) {
        dispatch({ type: actionTypes.success, payload })
        return payload
      } else {
        throw payload
      }
    } catch (error) {
      dispatch({ type: actionTypes.error, error })
      return error
    }
  }

  const handleSubmit = async (e) => {
    e.preventDefault()

    setErrors({})
    setIsLoading(true)
    const data = { email: email.value, password: password.value, confirm_password: confirmPassword.value }
    const response = await register(data)

    if (isSuccessResponse(response)) {
      const { from } = location.state || { from: { pathname: '/' } }
      setIsLoading(false)
      history.replace(from)
    } else {
      setErrors(response.data)
      setIsLoading(false)
    }
  }

  const showErrors = (errors) => {
    if (errors) {
      return Object.keys(errors).map((error) => (
        <p key={error} data-testid='register-error-item'>
          {errors[error].message}
        </p>
      ))
    }
  }

  return (
    <form onSubmit={handleSubmit} action='#'>
      <div>
        <input {...email} name='email' type='email' data-testid='register-input-email' />
        {showErrors(errors.email)}
      </div>

      <div>
        <input {...password} name='password' type='password' data-testid='register-input-password' />
        {showErrors(errors.password)}
      </div>

      <div>
        <input
          {...confirmPassword}
          name='confirm_password'
          type='password'
          data-testid='register-input-password-confirm'
        />
        {showErrors(errors.confirm_password)}
      </div>

      <div>
        <button type='submit' disabled={isLoading} data-testid='register-button-submit'>
          Register
        </button>
      </div>
    </form>
  )
}

export default Register
