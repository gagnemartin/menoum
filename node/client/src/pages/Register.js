import { useState } from 'react';
import { useHistory, useLocation } from 'react-router-dom'
import { useUserDispatch, register } from '../context/userContext'
import { isSuccessResponse } from '../global/helpers'

const Register = () => {
  const history = useHistory()
  const location = useLocation()
  const dispatch = useUserDispatch()

  const [email, setEmail] = useState('')
  const [password, setPassword] = useState('')
  const [confirmPassword, setConfirmPassword] = useState('')
  const [errors, setErrors] = useState({})
  const [isLoading, setIsLoading] = useState(false)

  const handleChange = (e) => {
    const {
      target: { name, value }
    } = e

    switch (name) {
      case 'email':
        setEmail(value)
        break
      case 'password':
        setPassword(value)
        break
      case 'confirm_password':
        setConfirmPassword(value)
        break
      default:
    }
  }

  const handleSubmit = async (e) => {
    e.preventDefault()

    setErrors({})
    setIsLoading(true)
    const data = { email, password, confirm_password: confirmPassword }
    const res = await register(dispatch, data)

    if (isSuccessResponse(res)) {
      const { from } = location.state || { from: { pathname: '/' } }
      setIsLoading(false)
      history.replace(from)
    } else {
      setErrors(res.data)
      setIsLoading(false)
    }
  }

  const showErrors = (errors) => {
    if (errors) {
      return Object.keys(errors).map((error) => <p key={error}>{errors[error].message}</p>)
    }
  }

  return (
    <form onSubmit={handleSubmit} action='#'>
      <div>
        <input onChange={handleChange} name='email' type='email' value={email} />
        {showErrors(errors.email)}
      </div>

      <div>
        <input onChange={handleChange} name='password' type='password' value={password} />
        {showErrors(errors.password)}
      </div>

      <div>
        <input onChange={handleChange} name='confirm_password' type='password' value={confirmPassword} />
        {showErrors(errors.confirm_password)}
      </div>

      <div>
        <button type='submit' disabled={isLoading}>
          Register
        </button>
      </div>
    </form>
  )
}

export default Register
