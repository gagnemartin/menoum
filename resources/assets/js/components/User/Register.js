import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Link, Redirect } from 'react-router-dom'

export default class Register extends Component
{
    constructor(props)
    {
        super(props)

        this.state = {
            username: null,
            email: null,
            password: null,
            password_confirmation: null
        }

        this.onSubmit = this.onSubmit.bind(this)
        this.handleChange = this.handleChange.bind(this)
    }

    handleChange(e)
    {
        const key = e.target.name
        let value = e.target.value

        if (value.trim() === '') {
            value = null
        }

        this.setState({
            [key]: value
        })
    }

    onSubmit(e)
    {
        e.preventDefault()

        const data = {
            username: this.state.username,
            email: this.state.email,
            password: this.state.password,
            password_confirmation: this.state.password_confirmation
        }

        this.props.onSubmit(data)
    }

    render()
    {
        const auth = this.props.auth

        return (
            auth.isAuthenticated ? <Redirect to="/" />
                :
                <div className="container">
                    <form className="form-basic user-register" method="POST" action="/api/register" onSubmit={ this.onSubmit }>
                        <h1 className="h3 mb-3 font-weight-normal">Login</h1>

                        <label htmlFor="username" className="sr-only">Username</label>
                        <input
                            onChange={ this.handleChange }
                            id="username"
                            type="text"
                            className="form-control"
                            name="username"
                            placeholder="Username"
                            required
                            autoFocus
                        />

                        <label htmlFor="email" className="sr-only">Email</label>
                        <input
                            onChange={ this.handleChange }
                            id="email"
                            type="email"
                            className="form-control"
                            name="email"
                            placeholder="Email"
                            required
                        />

                        <label htmlFor="password" className="sr-only">Password</label>
                        <input
                            onChange={ this.handleChange }
                            type="password"
                            name="password"
                            id="password"
                            className="form-control"
                            placeholder="Password"
                            required
                        />

                        <label htmlFor="passwordConfirmation" className="sr-only">Confirm your password</label>
                        <input
                            onChange={ this.handleChange }
                            type="password"
                            name="password_confirmation"
                            id="passwordConfirmation"
                            className="form-control"
                            placeholder="Confirm your password"
                            required
                        />

                        <button className="btn btn-lg btn-primary btn-block mb-2" type="submit">Register</button>
                        <Link to="/login" className="btn btn-lg btn-primary btn-block">Sign in</Link>
                        <Link to="/password-reset" className="btn btn-lg btn-primary btn-block">Forgot Your Password?</Link>
                    </form>
                </div>
        )
    }
}

Register.propTypes = {
    onSubmit: PropTypes.func.isRequired,
    auth: PropTypes.object.isRequired,
}

Register.defaultProps = {
    auth: {},
    onSubmit: () => {}
}