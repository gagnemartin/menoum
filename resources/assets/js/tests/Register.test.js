import React from 'react'
import { configure, shallow, mount, render } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import renderer from 'react-test-renderer'
import { MemoryRouter, Redirect } from 'react-router-dom'

import Register from '../components/User/Register'

configure({ adapter: new Adapter() })

describe('User register', () => {

    it("Can't access when authenticated", () => {
        const auth = {
            isAuthenticated: true
        }
        const component = shallow(<Register auth={ auth } />)
        const redirect = component.find(Redirect).exists()

        expect(redirect).toBeTruthy();
    })

    it('Matches the snapshot', () => {
        const component = renderer.create(<MemoryRouter><Register /></MemoryRouter>)
        const tree = component.toJSON()

        expect(tree).toMatchSnapshot()
    })

    it('Renders the form', () => {
        const component = shallow(<Register />)
        const $form = component.find('form.user-register')

        const form = $form.exists()
        expect(form).toBeTruthy()

        const usernameInput = $form.find('input[name="username"]').exists()
        expect(usernameInput).toBeTruthy()

        const emailInput = $form.find('input[name="email"]').exists()
        expect(emailInput).toBeTruthy()

        const passwordInput = $form.find('input[name="password"]').exists()
        expect(passwordInput).toBeTruthy()

        const passwordConfirmInput = $form.find('input[name="password_confirmation"]').exists()
        expect(passwordConfirmInput).toBeTruthy()
    })

    it('Changing input values changes the state', () => {
        const component = shallow(<Register />)
        const event = {
            target: {
                name: 'username',
                value: 'myUsername'
            }
        }

        const usernameInput = component.find('input[name="username"]')
        usernameInput.simulate('change', event)
        expect(component.state().username).toEqual(event.target.value)

        const emailInput = component.find('input[name="email"]')
        event.target.name = 'email'
        event.target.value = 'email@test.com'
        emailInput.simulate('change', event)
        expect(component.state().email).toEqual(event.target.value)

        const passwordInput = component.find('input[name="password"]')
        event.target.name = 'password'
        event.target.value = 'myPassword'
        passwordInput.simulate('change', event)
        expect(component.state().password).toEqual(event.target.value)

        const passwordConfirmInput = component.find('input[name="password_confirmation"]')
        event.target.name = 'password_confirmation'
        event.target.value = 'myPasswordConfirm'
        passwordConfirmInput.simulate('change', event)
        expect(component.state().password_confirmation).toEqual(event.target.value)
    })

    it('Submits the form', () => {
        const component = shallow(<Register />)
        const preventDefault = jest.fn()
        const form = component.find('form.user-register')

        form.simulate('submit', { preventDefault })
        expect(preventDefault).toHaveBeenCalled()
    })
})