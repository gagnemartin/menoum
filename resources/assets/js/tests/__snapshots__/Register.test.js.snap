// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User register Matches the snapshot 1`] = `
<div
  className="container"
>
  <form
    action="/api/register"
    className="form-basic user-register"
    method="POST"
    onSubmit={[Function]}
  >
    <h1
      className="h3 mb-3 font-weight-normal"
    >
      Login
    </h1>
    <label
      className="sr-only"
      htmlFor="username"
    >
      Username
    </label>
    <input
      autoFocus={true}
      className="form-control"
      id="username"
      name="username"
      onChange={[Function]}
      placeholder="Username"
      required={true}
      type="text"
    />
    <label
      className="sr-only"
      htmlFor="email"
    >
      Email
    </label>
    <input
      className="form-control"
      id="email"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      required={true}
      type="email"
    />
    <label
      className="sr-only"
      htmlFor="password"
    >
      Password
    </label>
    <input
      className="form-control"
      id="password"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      required={true}
      type="password"
    />
    <label
      className="sr-only"
      htmlFor="passwordConfirmation"
    >
      Confirm your password
    </label>
    <input
      className="form-control"
      id="passwordConfirmation"
      name="password_confirmation"
      onChange={[Function]}
      placeholder="Confirm your password"
      required={true}
      type="password"
    />
    <button
      className="btn btn-lg btn-primary btn-block mb-2"
      type="submit"
    >
      Register
    </button>
    <a
      className="btn btn-lg btn-primary btn-block"
      href="/login"
      onClick={[Function]}
    >
      Sign in
    </a>
    <a
      className="btn btn-lg btn-primary btn-block"
      href="/password-reset"
      onClick={[Function]}
    >
      Forgot Your Password?
    </a>
  </form>
</div>
`;
